File: SOLID-Principles/reference-application/Java/src/test/java/com/bounteous/ecom/EcommerceManagerTests.java
package com.bounteous.ecom;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.AfterEach;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.List;
import java.util.UUID;

import static org.assertj.core.api.Assertions.*;

class EcommerceManagerTests {

    private EcommerceManager ecommerceManager;
    private static final String CONNECTION_STRING = "jdbc:postgresql://localhost:5432/bounteous_ecom";
    private static final String USERNAME = "postgres";
    private static final String PASSWORD = "postgres123";

    @BeforeEach
    void setUp() {
        ecommerceManager = new EcommerceManager();
    }

    @AfterEach
    void tearDown() {
        // Clean up test data after each test (except products table which is seeded)
        cleanupTestData();
    }

    private void cleanupTestData() {
        try (Connection connection = DriverManager.getConnection(CONNECTION_STRING, USERNAME, PASSWORD);
             Statement statement = connection.createStatement()) {
            
            // Clean up tables in reverse dependency order
            statement.executeUpdate("DELETE FROM invoices");
            statement.executeUpdate("DELETE FROM order_items");
            statement.executeUpdate("DELETE FROM orders");
            statement.executeUpdate("DELETE FROM cart_items");
            statement.executeUpdate("DELETE FROM carts");
            statement.executeUpdate("DELETE FROM customers");
            
        } catch (SQLException ex) {
            // Log cleanup errors but don't fail tests
            System.err.println("Cleanup error: " + ex.getMessage());
        }
    }

    @Test
    void getProducts_ShouldReturnListOfProducts() throws SQLException {
        // Act
        List<Product> result = ecommerceManager.getProducts();

        // Assert
        assertThat(result).isNotNull();
        assertThat(result).isInstanceOf(List.class);
        // Should have sample products from database initialization
        assertThat(result).hasSizeGreaterThan(0);
    }

    @Test
    void createCustomer_ShouldReturnCustomerId() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        String firstName = "John";
        String lastName = "Doe";
        String phone = "555-0101";
        String address = "123 Main St";

        // Act
        int result = ecommerceManager.createCustomer(email, firstName, lastName, phone, address);

        // Assert
        assertThat(result).isGreaterThan(0);
        
        // Verify customer was actually created by retrieving it
        Customer customer = ecommerceManager.getCustomer(result);
        assertThat(customer).isNotNull();
        assertThat(customer.getEmail()).isEqualTo(email);
        assertThat(customer.getFirstName()).isEqualTo(firstName);
        assertThat(customer.getLastName()).isEqualTo(lastName);
    }

    @Test
    void createCustomer_WithNullEmail_ShouldThrowException() {
        // Arrange
        String email = null;
        String firstName = "John";
        String lastName = "Doe";
        String phone = "555-0101";
        String address = "123 Main St";

        // Act & Assert
        assertThatThrownBy(() -> ecommerceManager.createCustomer(email, firstName, lastName, phone, address))
                .isInstanceOf(Exception.class)
                .hasMessageContaining("null value in column \"email\"");
    }

    @Test
    void createCustomer_WithNullFirstName_ShouldThrowException() {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        String firstName = null;
        String lastName = "Doe";
        String phone = "555-0101";
        String address = "123 Main St";

        // Act & Assert
        assertThatThrownBy(() -> ecommerceManager.createCustomer(email, firstName, lastName, phone, address))
                .isInstanceOf(Exception.class)
                .hasMessageContaining("null value in column \"first_name\"");
    }

    @Test
    void createCustomer_WithNullLastName_ShouldThrowException() {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        String firstName = "John";
        String lastName = null;
        String phone = "555-0101";
        String address = "123 Main St";

        // Act & Assert
        assertThatThrownBy(() -> ecommerceManager.createCustomer(email, firstName, lastName, phone, address))
                .isInstanceOf(Exception.class)
                .hasMessageContaining("null value in column \"last_name\"");
    }

    @Test
    void createCustomer_WithDuplicateEmail_ShouldThrowException() throws SQLException {
        // Arrange
        String email = "duplicate" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        String firstName = "John";
        String lastName = "Doe";
        String phone = "555-0101";
        String address = "123 Main St";

        // Create first customer
        ecommerceManager.createCustomer(email, firstName, lastName, phone, address);

        // Act & Assert - Try to create second customer with same email
        assertThatThrownBy(() -> ecommerceManager.createCustomer(email, "Jane", "Smith", "555-0102", "456 Oak Ave"))
                .isInstanceOf(Exception.class)
                .hasMessageContaining("duplicate key value violates unique constraint");
    }

    @Test
    void createCart_ShouldReturnCartId() throws SQLException {
        // Arrange - First create a customer
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Jane", "Smith", "555-0102", "456 Oak Ave");

        // Act
        int result = ecommerceManager.createCart(customerId);

        // Assert
        assertThat(result).isGreaterThan(0);
        
        // Verify cart was actually created by retrieving it
        Cart cart = ecommerceManager.getCartByCustomerId(customerId);
        assertThat(cart).isNotNull();
        assertThat(cart.getId()).isEqualTo(result);
        assertThat(cart.getCustomerId()).isEqualTo(customerId);
    }

    @Test
    void createCart_WithInvalidCustomerId_ShouldThrowException() {
        // Arrange
        int invalidCustomerId = 99999; // Non-existent customer

        // Act & Assert
        assertThatThrownBy(() -> ecommerceManager.createCart(invalidCustomerId))
                .isInstanceOf(Exception.class)
                .hasMessageContaining("violates foreign key constraint");
    }

    @Test
    void addToCart_WithValidCustomerAndProduct_ShouldSucceed() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Test", "User", "555-0103", "789 Pine St");
        int cartId = ecommerceManager.createCart(customerId);
        List<Product> products = ecommerceManager.getProducts();
        int productId = products.get(0).getId(); // Use first available product
        int quantity = 2;

        // Act
        ecommerceManager.addToCart(customerId, productId, quantity);

        // Assert - Verify item was added by checking cart total
        double cartTotal = ecommerceManager.calculateCartTotal(customerId);
        assertThat(cartTotal).isGreaterThan(0);
        
        // Verify cart contents
        Cart cart = ecommerceManager.getCartByCustomerId(customerId);
        assertThat(cart.getItems()).hasSize(1);
        assertThat(cart.getItems().get(0).getProductId()).isEqualTo(productId);
        assertThat(cart.getItems().get(0).getQuantity()).isEqualTo(quantity);
    }

    @Test
    void addToCart_WithInvalidCustomer_ShouldNotThrow() throws SQLException {
        // Arrange
        int customerId = 99999; // Non-existent customer
        int productId = 1;
        int quantity = 2;

        // Act & Assert
        assertThatCode(() -> ecommerceManager.addToCart(customerId, productId, quantity))
                .doesNotThrowAnyException();
        
        // Verify no cart was created for invalid customer
        Cart cart = ecommerceManager.getCartByCustomerId(customerId);
        assertThat(cart).isNull();
    }

    @Test
    void addToCart_WithInvalidProductId_ShouldNotThrow() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Test", "User", "555-0103", "789 Pine St");
        int cartId = ecommerceManager.createCart(customerId);
        int invalidProductId = 99999; // Non-existent product
        int quantity = 2;

        // Act & Assert - The method should not throw because it validates product existence first
        assertThatCode(() -> ecommerceManager.addToCart(customerId, invalidProductId, quantity))
                .doesNotThrowAnyException();
    }

    @Test
    void addToCart_WithZeroQuantity_ShouldNotThrow() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Test", "User", "555-0103", "789 Pine St");
        int cartId = ecommerceManager.createCart(customerId);
        List<Product> products = ecommerceManager.getProducts();
        int productId = products.get(0).getId();
        int quantity = 0; // Zero quantity

        // Act & Assert - The method should not throw because it validates product existence first
        assertThatCode(() -> ecommerceManager.addToCart(customerId, productId, quantity))
                .doesNotThrowAnyException();
    }

    @Test
    void calculateCartTotal_WithEmptyCart_ShouldReturnZero() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Empty", "Cart", "555-0104", "321 Elm St");
        int cartId = ecommerceManager.createCart(customerId);

        // Act
        double result = ecommerceManager.calculateCartTotal(customerId);

        // Assert
        assertThat(result).isEqualTo(0.0);
    }

    @Test
    void calculateCartTotal_WithItems_ShouldReturnCorrectTotal() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Cart", "Test", "555-0105", "654 Maple St");
        int cartId = ecommerceManager.createCart(customerId);
        List<Product> products = ecommerceManager.getProducts();
        int productId = products.get(0).getId();
        int quantity = 2;
        double expectedTotal = products.get(0).getPrice() * quantity;

        ecommerceManager.addToCart(customerId, productId, quantity);

        // Act
        double result = ecommerceManager.calculateCartTotal(customerId);

        // Assert
        assertThat(result).isEqualTo(expectedTotal);
    }

    @Test
    void processOrder_WithEmptyCart_ShouldNotThrow() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Order", "Test", "555-0106", "987 Cedar St");
        int cartId = ecommerceManager.createCart(customerId);

        // Act & Assert
        assertThatCode(() -> ecommerceManager.processOrder(customerId))
                .doesNotThrowAnyException();
    }

    @Test
    void processOrder_WithItems_ShouldCreateOrderAndInvoice() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Order", "Process", "555-0107", "147 Birch St");
        int cartId = ecommerceManager.createCart(customerId);
        List<Product> products = ecommerceManager.getProducts();
        int productId = products.get(0).getId();
        int quantity = 1;

        ecommerceManager.addToCart(customerId, productId, quantity);

        // Act
        ecommerceManager.processOrder(customerId);

        // Assert - Verify cart is cleared after order processing
        double cartTotal = ecommerceManager.calculateCartTotal(customerId);
        assertThat(cartTotal).isEqualTo(0);
    }

    @Test
    void updateCustomer_ShouldUpdateCustomerData() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Original", "Name", "555-0108", "258 Spruce St");
        
        String newEmail = "updated" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        String newFirstName = "Updated";
        String newLastName = "Name";
        String newPhone = "555-0109";
        String newAddress = "369 Willow St";

        // Act
        ecommerceManager.updateCustomer(customerId, newEmail, newFirstName, newLastName, newPhone, newAddress);

        // Assert
        Customer updatedCustomer = ecommerceManager.getCustomer(customerId);
        assertThat(updatedCustomer).isNotNull();
        assertThat(updatedCustomer.getEmail()).isEqualTo(newEmail);
        assertThat(updatedCustomer.getFirstName()).isEqualTo(newFirstName);
        assertThat(updatedCustomer.getLastName()).isEqualTo(newLastName);
        assertThat(updatedCustomer.getPhone()).isEqualTo(newPhone);
        assertThat(updatedCustomer.getAddress()).isEqualTo(newAddress);
    }

    @Test
    void deleteCustomer_ShouldRemoveCustomer() throws SQLException {
        // Arrange
        String email = "test" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Delete", "Test", "555-0110", "741 Poplar St");

        // Verify customer exists
        Customer customer = ecommerceManager.getCustomer(customerId);
        assertThat(customer).isNotNull();

        // Act
        ecommerceManager.deleteCustomer(customerId);

        // Assert
        Customer deletedCustomer = ecommerceManager.getCustomer(customerId);
        assertThat(deletedCustomer).isNull();
    }

    @Test
    void generateSalesReport_ShouldNotThrow() throws SQLException {
        // Act & Assert
        assertThatCode(() -> ecommerceManager.generateSalesReport())
                .doesNotThrowAnyException();
    }

    @Test
    void updateProductStock_ShouldNotThrow() throws SQLException {
        // Arrange
        List<Product> products = ecommerceManager.getProducts();
        int productId = products.get(0).getId();
        int newStock = 100;

        // Act & Assert
        assertThatCode(() -> ecommerceManager.updateProductStock(productId, newStock))
                .doesNotThrowAnyException();
    }

    @Test
    void product_ShouldViolateOpenClosedPrinciple() {
        // This test demonstrates that the Product class violates OCP
        // by being hard-coded and not easily extensible

        Product product = new Product(1, "Test Product", 99.99, "TEST-001");

        assertThat(product).isNotNull();
        assertThat(product.getId()).isEqualTo(1);
        assertThat(product.getName()).isEqualTo("Test Product");
        assertThat(product.getPrice()).isEqualTo(99.99);
        assertThat(product.getSku()).isEqualTo("TEST-001");

        // The Product class is hard-coded and cannot be extended without modification
        // This violates the Open/Closed Principle
    }

    @Test
    void customer_ShouldViolateOpenClosedPrinciple() {
        // This test demonstrates that the Customer class violates OCP
        // by being hard-coded and not easily extensible

        Customer customer = new Customer(1, "test@example.com", "John", "Doe", "555-0101", "123 Main St");

        assertThat(customer).isNotNull();
        assertThat(customer.getId()).isEqualTo(1);
        assertThat(customer.getEmail()).isEqualTo("test@example.com");
        assertThat(customer.getFirstName()).isEqualTo("John");
        assertThat(customer.getLastName()).isEqualTo("Doe");
        assertThat(customer.getPhone()).isEqualTo("555-0101");
        assertThat(customer.getAddress()).isEqualTo("123 Main St");

        // The Customer class is hard-coded and cannot be extended without modification
        // This violates the Open/Closed Principle
    }

    @Test
    void cart_ShouldViolateOpenClosedPrinciple() {
        // This test demonstrates that the Cart class violates OCP
        // by being hard-coded and not easily extensible

        Cart cart = new Cart(1, 1, List.of());

        assertThat(cart).isNotNull();
        assertThat(cart.getId()).isEqualTo(1);
        assertThat(cart.getCustomerId()).isEqualTo(1);
        assertThat(cart.getItems()).isEmpty();

        // The Cart class is hard-coded and cannot be extended without modification
        // This violates the Open/Closed Principle
    }

    @Test
    void order_ShouldViolateOpenClosedPrinciple() {
        // This test demonstrates that the Order class violates OCP
        // by being hard-coded and not easily extensible

        Order order = new Order();
        order.setId(1);
        order.setCustomerId(1);
        order.setOrderNumber("ORD-001");

        assertThat(order).isNotNull();
        assertThat(order.getId()).isEqualTo(1);
        assertThat(order.getCustomerId()).isEqualTo(1);
        assertThat(order.getOrderNumber()).isEqualTo("ORD-001");

        // The Order class is hard-coded and cannot be extended without modification
        // This violates the Open/Closed Principle
    }

    @Test
    void invoice_ShouldViolateOpenClosedPrinciple() {
        // This test demonstrates that the Invoice class violates OCP
        // by being hard-coded and not easily extensible

        Invoice invoice = new Invoice();
        invoice.setId(1);
        invoice.setOrderId(1);
        invoice.setInvoiceNumber("INV-001");

        assertThat(invoice).isNotNull();
        assertThat(invoice.getId()).isEqualTo(1);
        assertThat(invoice.getOrderId()).isEqualTo(1);
        assertThat(invoice.getInvoiceNumber()).isEqualTo("INV-001");

        // The Invoice class is hard-coded and cannot be extended without modification
        // This violates the Open/Closed Principle
    }

    @Test
    void cartWorkflow_ShouldDemonstrateMultipleSOLIDViolations() throws SQLException {
        // This test demonstrates multiple SOLID principle violations in a real workflow
        
        // Arrange
        String email = "workflow" + UUID.randomUUID().toString().substring(0, 8) + "@example.com";
        int customerId = ecommerceManager.createCustomer(email, "Workflow", "Test", "555-0111", "852 Ash St");
        
        // Act - Complete cart workflow
        int cartId = ecommerceManager.createCart(customerId);
        List<Product> products = ecommerceManager.getProducts();
        int productId = products.get(0).getId();
        
        ecommerceManager.addToCart(customerId, productId, 2);
        double cartTotal = ecommerceManager.calculateCartTotal(customerId);
        ecommerceManager.processOrder(customerId);
        
        // Assert - Verify the workflow completed successfully
        assertThat(cartId).isGreaterThan(0);
        assertThat(cartTotal).isGreaterThan(0);
        
        // This workflow demonstrates violations of:
        // - SRP: EcommerceManager handles multiple responsibilities
        // - OCP: Hard-coded classes that can't be extended
        // - DIP: Direct database dependencies
    }
}

File: MarkDown/1-SOLID-Principles/backend/reference-application/Dotnet/src/App.Tests/EcommerceManagerTests.cs
using Xunit;
using FluentAssertions;
using System;
using System.Collections.Generic;
using EcomApp;
using Npgsql;

namespace App.Tests
{
    public class EcommerceManagerTests : IDisposable
    {
        private readonly EcommerceManager _ecommerceManager;
        private readonly string _connectionString;

        public EcommerceManagerTests()
        {
            _ecommerceManager = new EcommerceManager();
            _connectionString = "Host=localhost;Database=bounteous_ecom;Username=postgres;Password=postgres123";
        }

        public void Dispose()
        {
            // Clean up test data after each test (except products table which is seeded)
            CleanupTestData();
        }

        private void CleanupTestData()
        {
            try
            {
                using var connection = new NpgsqlConnection(_connectionString);
                connection.Open();
                
                // Clean up tables in reverse dependency order
                using var command = new NpgsqlCommand(@"
                    DELETE FROM invoices;
                    DELETE FROM order_items;
                    DELETE FROM orders;
                    DELETE FROM cart_items;
                    DELETE FROM carts;
                    DELETE FROM customers;
                ", connection);
                
                command.ExecuteNonQuery();
            }
            catch (Exception ex)
            {
                // Log cleanup errors but don't fail tests
                Console.WriteLine($"Cleanup error: {ex.Message}");
            }
        }

        [Fact]
        public void GetProducts_ShouldReturnListOfProducts()
        {
            // Act
            var result = _ecommerceManager.GetProducts();

            // Assert
            result.Should().NotBeNull();
            result.Should().BeOfType<List<Product>>();
            // Should have sample products from database initialization
            result.Should().HaveCountGreaterThan(0);
        }

        [Fact]
        public void CreateCustomer_ShouldReturnCustomerId()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var firstName = "John";
            var lastName = "Doe";
            var phone = "555-0101";
            var address = "123 Main St";

            // Act
            var result = _ecommerceManager.CreateCustomer(email, firstName, lastName, phone, address);

            // Assert
            result.Should().BeGreaterThan(0);
            
            // Verify customer was actually created by retrieving it
            var customer = _ecommerceManager.GetCustomer(result);
            customer.Should().NotBeNull();
            customer.Email.Should().Be(email);
            customer.FirstName.Should().Be(firstName);
            customer.LastName.Should().Be(lastName);
        }

        [Fact]
        public void CreateCustomer_WithNullEmail_ShouldThrowException()
        {
            // Arrange
            string email = null;
            var firstName = "John";
            var lastName = "Doe";
            var phone = "555-0101";
            var address = "123 Main St";

            // Act & Assert
            var action = () => _ecommerceManager.CreateCustomer(email, firstName, lastName, phone, address);
            action.Should().Throw<Exception>("Expected exception for null email constraint violation");
        }

        [Fact]
        public void CreateCustomer_WithNullFirstName_ShouldThrowException()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            string firstName = null;
            var lastName = "Doe";
            var phone = "555-0101";
            var address = "123 Main St";

            // Act & Assert
            var action = () => _ecommerceManager.CreateCustomer(email, firstName, lastName, phone, address);
            action.Should().Throw<Exception>("Expected exception for null first_name constraint violation");
        }

        [Fact]
        public void CreateCustomer_WithNullLastName_ShouldThrowException()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var firstName = "John";
            string lastName = null;
            var phone = "555-0101";
            var address = "123 Main St";

            // Act & Assert
            var action = () => _ecommerceManager.CreateCustomer(email, firstName, lastName, phone, address);
            action.Should().Throw<Exception>("Expected exception for null last_name constraint violation");
        }

        [Fact]
        public void CreateCustomer_WithDuplicateEmail_ShouldThrowException()
        {
            // Arrange
            var email = $"duplicate{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var firstName = "John";
            var lastName = "Doe";
            var phone = "555-0101";
            var address = "123 Main St";

            // Create first customer
            _ecommerceManager.CreateCustomer(email, firstName, lastName, phone, address);

            // Act & Assert - Try to create second customer with same email
            var action = () => _ecommerceManager.CreateCustomer(email, "Jane", "Smith", "555-0102", "456 Oak Ave");
            action.Should().Throw<Exception>("Expected exception for duplicate email constraint violation");
        }

        [Fact]
        public void CreateCart_ShouldReturnCartId()
        {
            // Arrange - First create a customer
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Jane", "Smith", "555-0102", "456 Oak Ave");

            // Act
            var result = _ecommerceManager.CreateCart(customerId);

            // Assert
            result.Should().BeGreaterThan(0);
            
            // Verify cart was actually created by retrieving it
            var cart = _ecommerceManager.GetCartByCustomerId(customerId);
            cart.Should().NotBeNull();
            cart.Id.Should().Be(result);
            cart.CustomerId.Should().Be(customerId);
        }

        [Fact]
        public void CreateCart_WithInvalidCustomerId_ShouldThrowException()
        {
            // Arrange
            var invalidCustomerId = 99999; // Non-existent customer

            // Act & Assert
            var action = () => _ecommerceManager.CreateCart(invalidCustomerId);
            action.Should().Throw<Exception>("Expected exception for foreign key constraint violation on customer_id");
        }

        [Fact]
        public void AddToCart_WithValidCustomerAndProduct_ShouldSucceed()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Test", "User", "555-0103", "789 Pine St");
            var cartId = _ecommerceManager.CreateCart(customerId);
            var products = _ecommerceManager.GetProducts();
            var productId = products[0].Id; // Use first available product
            var quantity = 2;

            // Act
            _ecommerceManager.AddToCart(customerId, productId, quantity);

            // Assert - Verify item was added by checking cart total
            var cartTotal = _ecommerceManager.CalculateCartTotal(customerId);
            cartTotal.Should().BeGreaterThan(0);
            
            // Verify cart contents
            var cart = _ecommerceManager.GetCartByCustomerId(customerId);
            cart.Items.Should().HaveCount(1);
            cart.Items[0].ProductId.Should().Be(productId);
            cart.Items[0].Quantity.Should().Be(quantity);
        }

        [Fact]
        public void AddToCart_WithInvalidCustomer_ShouldNotThrow()
        {
            // Arrange
            var customerId = 99999; // Non-existent customer
            var productId = 1;
            var quantity = 2;

            // Act & Assert
            var action = () => _ecommerceManager.AddToCart(customerId, productId, quantity);
            action.Should().NotThrow();
            
            // Verify no cart was created for invalid customer
            var cart = _ecommerceManager.GetCartByCustomerId(customerId);
            cart.Should().BeNull();
        }

        [Fact]
        public void AddToCart_WithInvalidProductId_ShouldNotThrow()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Test", "User", "555-0103", "789 Pine St");
            var cartId = _ecommerceManager.CreateCart(customerId);
            var invalidProductId = 99999; // Non-existent product
            var quantity = 2;

            // Act & Assert - The method should not throw because it validates product existence first
            var action = () => _ecommerceManager.AddToCart(customerId, invalidProductId, quantity);
            action.Should().NotThrow("The method validates product existence and returns early without database operation");
        }

        [Fact]
        public void AddToCart_WithZeroQuantity_ShouldNotThrow()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Test", "User", "555-0103", "789 Pine St");
            var cartId = _ecommerceManager.CreateCart(customerId);
            var products = _ecommerceManager.GetProducts();
            var productId = products[0].Id;
            var quantity = 0; // Zero quantity

            // Act & Assert - The method should not throw because it validates product existence first
            var action = () => _ecommerceManager.AddToCart(customerId, productId, quantity);
            action.Should().NotThrow("The method validates product existence and returns early without database operation");
        }

        [Fact]
        public void CalculateCartTotal_WithEmptyCart_ShouldReturnZero()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Empty", "Cart", "555-0104", "321 Elm St");
            var cartId = _ecommerceManager.CreateCart(customerId);

            // Act
            var result = _ecommerceManager.CalculateCartTotal(customerId);

            // Assert
            result.Should().Be(0);
        }

        [Fact]
        public void CalculateCartTotal_WithItems_ShouldReturnCorrectTotal()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Cart", "Test", "555-0105", "654 Maple St");
            var cartId = _ecommerceManager.CreateCart(customerId);
            var products = _ecommerceManager.GetProducts();
            var productId = products[0].Id;
            var quantity = 2;
            var expectedTotal = products[0].Price * quantity;

            _ecommerceManager.AddToCart(customerId, productId, quantity);

            // Act
            var result = _ecommerceManager.CalculateCartTotal(customerId);

            // Assert
            result.Should().Be(expectedTotal);
        }

        [Fact]
        public void ProcessOrder_WithEmptyCart_ShouldNotThrow()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Order", "Test", "555-0106", "987 Cedar St");
            var cartId = _ecommerceManager.CreateCart(customerId);

            // Act & Assert
            var action = () => _ecommerceManager.ProcessOrder(customerId);
            action.Should().NotThrow();
        }

        [Fact]
        public void ProcessOrder_WithItems_ShouldCreateOrderAndInvoice()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Order", "Process", "555-0107", "147 Birch St");
            var cartId = _ecommerceManager.CreateCart(customerId);
            var products = _ecommerceManager.GetProducts();
            var productId = products[0].Id;
            var quantity = 1;

            _ecommerceManager.AddToCart(customerId, productId, quantity);

            // Act
            _ecommerceManager.ProcessOrder(customerId);

            // Assert - Verify cart is cleared after order processing
            var cartTotal = _ecommerceManager.CalculateCartTotal(customerId);
            cartTotal.Should().Be(0);
        }

        [Fact]
        public void UpdateCustomer_ShouldUpdateCustomerData()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Original", "Name", "555-0108", "258 Spruce St");
            
            var newEmail = $"updated{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var newFirstName = "Updated";
            var newLastName = "Name";
            var newPhone = "555-0109";
            var newAddress = "369 Willow St";

            // Act
            _ecommerceManager.UpdateCustomer(customerId, newEmail, newFirstName, newLastName, newPhone, newAddress);

            // Assert
            var updatedCustomer = _ecommerceManager.GetCustomer(customerId);
            updatedCustomer.Should().NotBeNull();
            updatedCustomer.Email.Should().Be(newEmail);
            updatedCustomer.FirstName.Should().Be(newFirstName);
            updatedCustomer.LastName.Should().Be(newLastName);
            updatedCustomer.Phone.Should().Be(newPhone);
            updatedCustomer.Address.Should().Be(newAddress);
        }

        [Fact]
        public void DeleteCustomer_ShouldRemoveCustomer()
        {
            // Arrange
            var email = $"test{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Delete", "Test", "555-0110", "741 Poplar St");

            // Verify customer exists
            var customer = _ecommerceManager.GetCustomer(customerId);
            customer.Should().NotBeNull();

            // Act
            _ecommerceManager.DeleteCustomer(customerId);

            // Assert
            var deletedCustomer = _ecommerceManager.GetCustomer(customerId);
            deletedCustomer.Should().BeNull();
        }

        [Fact]
        public void GenerateSalesReport_ShouldNotThrow()
        {
            // Act & Assert
            var action = () => _ecommerceManager.GenerateSalesReport();
            action.Should().NotThrow();
        }

        [Fact]
        public void UpdateProductStock_ShouldNotThrow()
        {
            // Arrange
            var products = _ecommerceManager.GetProducts();
            var productId = products[0].Id;
            var newStock = 100;

            // Act & Assert
            var action = () => _ecommerceManager.UpdateProductStock(productId, newStock);
            action.Should().NotThrow();
        }

        [Fact]
        public void EcommerceManager_ShouldViolateSingleResponsibilityPrinciple()
        {
            // This test demonstrates that the EcommerceManager class violates SRP
            // by having multiple responsibilities

            // Assert - The class handles multiple concerns
            _ecommerceManager.Should().NotBeNull();
            
            // The class should have methods for:
            // - Product management
            // - Customer management  
            // - Cart management
            // - Order processing
            // - Invoice generation
            // - Email sending
            // - Logging
            // - Inventory management
            // - Reporting
            
            // This is a violation of SRP as the class has too many reasons to change
        }

        [Fact]
        public void Product_ShouldViolateOpenClosedPrinciple()
        {
            // This test demonstrates that the Product class violates OCP
            // by being hard-coded and not easily extensible

            var product = new Product { Id = 1, Name = "Test Product", Price = 99.99m, Sku = "TEST-001" };
            
            product.Should().NotBeNull();
            product.Id.Should().Be(1);
            product.Name.Should().Be("Test Product");
            product.Price.Should().Be(99.99m);
            product.Sku.Should().Be("TEST-001");
            
            // The Product class is hard-coded and cannot be extended without modification
            // This violates the Open/Closed Principle
        }

        [Fact]
        public void Customer_ShouldViolateOpenClosedPrinciple()
        {
            // This test demonstrates that the Customer class violates OCP
            // by being hard-coded and not easily extensible

            var customer = new Customer 
            { 
                Id = 1, 
                Email = "test@example.com", 
                FirstName = "John", 
                LastName = "Doe", 
                Phone = "555-0101", 
                Address = "123 Main St" 
            };
            
            customer.Should().NotBeNull();
            customer.Id.Should().Be(1);
            customer.Email.Should().Be("test@example.com");
            customer.FirstName.Should().Be("John");
            customer.LastName.Should().Be("Doe");
            customer.Phone.Should().Be("555-0101");
            customer.Address.Should().Be("123 Main St");
            
            // The Customer class is hard-coded and cannot be extended without modification
            // This violates the Open/Closed Principle
        }

        [Fact]
        public void CartWorkflow_ShouldDemonstrateMultipleSOLIDViolations()
        {
            // This test demonstrates multiple SOLID principle violations in a real workflow
            
            // Arrange
            var email = $"workflow{Guid.NewGuid().ToString("N")[..8]}@example.com";
            var customerId = _ecommerceManager.CreateCustomer(email, "Workflow", "Test", "555-0111", "852 Ash St");
            
            // Act - Complete cart workflow
            var cartId = _ecommerceManager.CreateCart(customerId);
            var products = _ecommerceManager.GetProducts();
            var productId = products[0].Id;
            
            _ecommerceManager.AddToCart(customerId, productId, 2);
            var cartTotal = _ecommerceManager.CalculateCartTotal(customerId);
            _ecommerceManager.ProcessOrder(customerId);
            
            // Assert - Verify the workflow completed successfully
            cartId.Should().BeGreaterThan(0);
            cartTotal.Should().BeGreaterThan(0);
            
            // This workflow demonstrates violations of:
            // - SRP: EcommerceManager handles multiple responsibilities
            // - OCP: Hard-coded classes that can't be extended
            // - DIP: Direct database dependencies
        }
    }
}
